Dockerfile

```diff
@@ -1,23 +1,29 @@
-# Build the manager binary
-FROM golang:1.13.3 as builder
+FROM golang:1.14-alpine as builder
+
+ARG TARGETPLATFORM
+
+ENV GO111MODULE=on \
+    CGO_ENABLED=0
+
+RUN apk add --no-cache git
 
-# Copy in the go src
 WORKDIR /go/src/github.com/open-policy-agent/gatekeeper
-COPY pkg/    pkg/
-COPY third_party/ third_party/
-COPY vendor/ vendor/
-COPY main.go main.go
-COPY api/ api/
-COPY go.mod .
 
-# Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -mod vendor -a -o manager main.go
+COPY . .
```

Comment:
Instead of copying everything and relying on the .Dockerignore file, could we use the multi-source form of the copy statement?

https://docs.docker.com/engine/reference/builder/#copy

```
COPY pkg third_party vendor main.go api go.mod .
```

That way we still get only one layer and we remain explicit as to what is being copied?

I definitely appreciate the fewer layers.

Reply:
The following from that linked doc suggests that the directory names might not be preserved with the above approach:

> Note: The directory itself is not copied, just its contents.

If its a choice between explicitness and layer reduction, I'd prefer the explicitness rather than the layer reduction, as we shouldn't be uploading the build image anyway.

Reply:
That is reasonable, do we want to revert the .dockerignore, or leave it with these changes and just have them double fortify the things we COPY into the image at build time?

Reply:
Let's remove the .Dockerignore filter except for the files that should just never make it in (looks like line 12 onward, and maybe we keep ignoring .git).

Reply:
The directory structure comment is correct - You can always tweak either the dir structure and or use the ADD command and tarball the dirs into the image, although I think just having multiple layers will be acceptable - I personally like the .Dockerignore, perhaps I could have added a comment with a link above the `COPY . .` line, but I'm easy with both approaches.

Reply:
The .dockerignore approach couples the two files in both directions. In order to understand the dockerfile, you must know about how it leverages .dockerignore. In order to verify your change to .dockerignore isn't breaking, you must understand the dockerfile. I'm leaning toward explicitness here because it makes things less brittle and more readily verifiable.

Tar-ing + adding would work, but is an extra step. B/c the extra layers don't seem to add much functional overhead, I'd prefer keeping them.

Reply:
Excellent - happy with both approaches 👍